### PR TITLE
feat: handle distroless for upgrade all packages

### LINKED ISSUE #569: For distroless, copa might need to get list of packages first, since we don't have the package manager tooling.

### PR DESCRIPTION
**Pull request by ashnamehrotra**
Describe the changes in this pull request using active verbs such as _Add_, _Remove_, _Replace_ ...

- Adds support for distroless images when updating all packages without scanner report. 
- Reads from `/var/lib/rpmmanifest/container-manifest-2` for rpm and `/var/lib/dpkg/status.d/` for debian to get list of packages.
- Verifies which of these packages is outdated and updates only those packages.
- Handles an edge case for mac when host OS is darwin and sets it to linux.
- Tests same set of images as with scan report with ignore-errors false

Closes #569 
Part of #522

### CODE REVIEW

File: pkg/pkgmgr/dpkg.go [hunk 1]
```diff
@@ -303,21 +368,61 @@ func (dm *dpkgManager) unpackAndMergeUpdates(ctx context.Context, updates unvers
...
+		}
+		downloadCmd = fmt.Sprintf(aptDownloadTemplate, strings.Join(pkgStrings, " "))
+	} else {
+		// only updated the outdated pacakges from packages.txt
+		downloadCmd = "xargs -a packages.txt -n 1 apt download --no-install-recommends"
 	}
-	downloadCmd := fmt.Sprintf(aptDownloadTemplate, strings.Join(pkgStrings, " "))
-	downloaded := updated.Dir(downloadPath).Run(llb.Shlex(downloadCmd), llb.WithProxy(utils.GetProxy())).Root()
+
+	downloadPath := "/var/cache/apt/archives"
```
Comment (sozercan) [inline pkg/pkgmgr/dpkg.go:418]: is this okay to move to const?

File: integration/fixtures/test-images-update-all.json [hunk 2]
```diff
@@ -41,6 +41,31 @@
         "description": "Valid dpkg/status, apt present, custom network config",
         "ignoreErrors": false
     },
+    {
```
Comment (sozercan) [inline integration/fixtures/test-images-update-all.json:44]: are we at parity with the existing list? if so, can we keep a single list now?

Comment (ashnamehrotra) [reply integration/fixtures/test-images-update-all.json:44]: not completely, this doesn't test `registry.k8s.io/kube-proxy:v1.27.2` since it is both distroless and has a custom dpkg/status.d file. If this file does not follow the standard convention, we cannot read from it to get list of packages. This also sets ignoreErrors to false for `docker.io/grafana/grafana-image-renderer:3.4.0` since we are no longer trying to patch the nonexistent libraries from scan report.

Comment (ashnamehrotra) [reply integration/fixtures/test-images-update-all.json]: fixed error and changed to use same test-images.json file. Runs these images with ignore-errors false

Comment (ashnamehrotra) [reply integration/fixtures/test-images-update-all.json]: Also added kube-proxy v1.26.1 in integration test to check for the case that libssl and libssl1.1 are both present. Tested locally and kube-proxy v1.27.9 with both libss1.1 and libssl3 present will break integration tests.

File: integration/fixtures/test-images.json [hunk 3]
```diff
@@ -49,12 +49,20 @@
         "description": "Custom dpkg/status.d with text names, no apt, libssl1.1",
         "ignoreErrors": false
     },
+    {
+        "image": "registry.k8s.io/kube-proxy",
+        "tag": "v1.26.1",
+        "digest": "sha256:1e4f13f5f5c215813fb9c9c6f56da1c0354363f2a69bd12732658f79d585864f",
+        "distro": "Custom Google Distroless",
+        "description": "Custom dpkg/status.d with text names, with both libssl1.1 and libss1",
```
Comment (sozercan) [inline integration/fixtures/test-images.json:57]:
```suggestion
        "description": "Custom dpkg/status.d with text names, with both libssl1.1 and libssl1",
```

File: website/docs/update-all.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+---
```
Comment (sozercan) [inline website/docs/update-all.md:1]: can we add this as part of quick start instead of a new page? we might need to refactor a bit

Comment (sozercan) [reply website/docs/update-all.md:1]: mention both options and users can choose

File: pkg/buildkit/buildkit.go [hunk 5]
```diff
@@ -60,7 +61,12 @@ func InitializeBuildkitConfig(ctx context.Context, c gwclient.Client, image stri
 
 // Extracts the bytes of the file denoted by `path` from the state `st`.
 func ExtractFileFromState(ctx context.Context, c gwclient.Client, st *llb.State, path string) ([]byte, error) {
-	def, err := st.Marshal(ctx)
+	platform := platforms.Normalize(platforms.DefaultSpec())
+	if platform.OS != "linux" {
```
Comment (sozercan) [inline pkg/buildkit/buildkit.go:65]: was this because of darwin issue?

File: pkg/patch/patch.go [hunk 6]
```diff
@@ -319,6 +330,16 @@ func getOSType(ctx context.Context, osreleaseBytes []byte) (string, error) {
 	}
 }
 
+func getOSVersion(ctx context.Context, osreleaseBytes []byte) (string, error) {
```
Comment (sozercan) [inline pkg/patch/patch.go:333]: can we add a unit test for this one

File: pkg/pkgmgr/dpkg.go [hunk 7]
```diff
@@ -196,6 +216,7 @@ func (dm *dpkgManager) probeDPKGStatus(ctx context.Context, toolImage string) er
                 elif [ -d "$DPKG_STATUS_FOLDER" ]; then
                     status="$DPKG_STATUS_IS_DIRECTORY"
                     ls -1 "$DPKG_STATUS_FOLDER" > "$RESULT_STATUSD_PATH"
+					mv "$DPKG_STATUS_FOLDER"/*  "$RESULTS_PATH"
```
Comment (sozercan) [inline pkg/pkgmgr/dpkg.go:219]: nit: for consistency, let's keep using spaces

File: pkg/pkgmgr/dpkg.go [hunk 8]
```diff
@@ -227,6 +273,29 @@ func (dm *dpkgManager) probeDPKGStatus(ctx context.Context, toolImage string) er
 	}
 }
 
+func getPackageInfo(file string) (string, string, error) {
```
Comment (sozercan) [inline pkg/pkgmgr/dpkg.go:276]: can you add a unit test for this too

**Review by sozercan — APPROVED** [review]
Thank you! LGTM
